Added tests for client.js

ericman314 opened the pull request:
## Description
- Modified `Keypad` constructor to reuse the code in `setCode` and `setCodeLength`, to enforce maximum length requirements
- Fixed typo in `Client` constructor
- Added default value in several class constructors: `(params = {})`. Figured this was OK since every property in `params` has a default value anyway.
- Minor edits in several other class files
- Created tests for several of the class files


## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
#196

alexanderson1993 commented on server/src/classes/client.js:
```diff
@@ -64,6 +70,7 @@ class Scanner {
     this.scanning = true;
   }
   cancelScan() {
+    // TODO: Should we set this.scanRequest to '' ?
```
I want to keep the scan request intact so the crew and staff can know what the cancelled request was.

```suggestion 
```

alexanderson1993 commented on server/src/classes/crew.js:
```diff
@@ -62,6 +66,7 @@ export default class Crew {
   constructor(params) {
     this.id = params.id || uuid.v4();
     this.class = "Crew";
+    // TODO: Are these really the default params we want?
```
Yes. There have to be some default params, and I want it to be pretty obvious that they are default. Also, the client code is set up so it is unlikely that the default params will ever be used in this way.

```suggestion
```

alexanderson1993 commented on server/src/classes/docking.js:
```diff
@@ -20,6 +20,7 @@ export default class DockingPort extends System {
     }
   }
   updateDockingPort({ name, type, clamps, compress, doors, image, docked }) {
+    // Suggestion: change to if(arguments[0].hasOwnProperty('name'))
```
Actually, this would not accomplish what I'm intending. It's possible that `name` will be either `null` or `undefined`. `.hasOwnProperty(x)` will return true if `x` is a falsey value, which is not the intended behavior. The only falsey value that should work is `""` (empty string)
```suggestion
```

alexanderson1993 commented on server/tests/classes/client.test.js:
```diff
@@ -0,0 +1,347 @@
+
+  describe('addCache', () => {
+    test.skip('should work', () => {
+      // TODO: What does this do?
```
The Client object is a reference to a connected client. Caches allow the server to tell the client large files that it should preload, such as movies, images, or sound effects. All it does is throw the URI of the file in an array and push that to the client with a subscription publish. The client then uses a `<link rel="preload">` to preload the asset

ericman314 replied:
I will add some tests for this but I can't get to right this moment.

alexanderson1993 replied:
It's fine. You can leave the `// TODO` statements and we can get to this later.

alexanderson1993 commented on server/tests/classes/client.test.js:
```diff
@@ -0,0 +1,347 @@
+      c.keypad.setCodeLength(5);
+      expect(c.keypad.codeLength).toBe(5);
+
+      // TODO: What if codeLength doesn't match code.length?
```
It looks like you adjusted the code so `codeLength` is set to the length of the code you enter, which is the ideal behavior. `codeLength`, on the other hand, is only for randomly generated codes. So I would ignore the length of `this.code` in this test.


```suggestion
```

alexanderson1993 commented on server/tests/classes/coolant.test.js:
```diff
@@ -0,0 +1,57 @@
+    });
+  });
+
+  describe('power', () => {
```
The getter and the setter override the `System` class, because the coolant system shouldn't have power requirements. Unit testing this aspect isn't necessary.

alexanderson1993 commented:
And don't worry about the Codacy quality check. I'm not super concerned about it.

ericman314 commented:
Thanks for reviewing. I'm new to this so I wasn't sure about a few things.

And I'll definitely work in a new branch next time, I just forgot to create one. I figured I could get this one merged quickly enough so it wouldn't matter.

I'll review your comments and make the necessary changes.

ericman314 commented:
The latest commit addresses your comments, except I haven't written tests yet for `addCache` and `removeCache`. I can get to those tonight, but otherwise it's ready to merge. If you want you can merge now and I'll write tests for the caches for the next PR.
